[OSDOCS-3974]: new node tuning topic for hosted control planes

lahinson opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS-<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch. --->

Version(s): 4.12
<!--- Specify the version or versions of OpenShift your PR applies to. -->

<!-- Version examples:
  * PR applies to all versions after a specific version (e.g. 4.8): 4.8+
  * PR applies to the in-development version (e.g. 4.12) and future versions: 4.12+
  * PR applies only to a specific single version (e.g. 4.10): 4.10
  * PR applies to multiple specific versions (e.g. 4.8-4.10): 4.8, 4.9, 4.10 --->

Issue: https://issues.redhat.com/browse/OSDOCS-3974
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->

Link to docs preview: 
- Configuring node tuning in a hosted cluster: https://51302--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/using-node-tuning-operator.html#node-tuning-hosted-cluster_node-tuning-operator
- Advanced node tuning for hosted clusters by setting kernel boot parameters: https://51302--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/using-node-tuning-operator.html#advanced-node-tuning-hosted-cluster_node-tuning-operator
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->

QE review:
- [x] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

<!-- NOTE:
Automatic preview functionality is currently only available for some branches. For PRs that update the rendered build in any way against branches that do not create an automated preview:
  * OpenShift documentation team members (core and aligned) must include a link to a locally generated preview.
  * External contributors can request a generated preview from the OpenShift documentation team. --->

Additional information: This PR has been peer reviewed and has received QE and SME approval.
<!--- Optional: Include additional context or expand the description here.--->

<!--- Next steps after opening your PR:

* Ask for review from the OpenShift docs team:
  - For community authors: Request a review by tagging @openshift/team-documentation in a GitHub comment.
  - For Red Hat associates:
    * For normal peer requests, add a comment that contains this text: /label peer-review-needed
    * For normal merge review requests, add a comment that contains this text: /label merge-review-needed
    * For urgent peer review requests, ping @peer-review-squad requesting a review in the #forum-docs-review channel (CoreOS Slack workspace) and provide the following information:
      * A link to the PR.
      * The size of the PR that the GitHub bot assigns (ex: XS, S, M, L, XL).
      * Details about how the PR is urgent.
    * For urgent merge requests, ping @merge-review-squad in the #forum-docs-review channel (CoreOS Slack workspace).
    * Except for changes that do not impact the meaning of the content, QE review is required before content is merged.

* IMPORTANT:
  - All documentation changes must be verified by a QE team associate before merging.
  - Squash to one commit before submitting your PR for peer review.

* For more information about verifying your content, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#verification-of-your-content

* For more information about contributing to OpenShift documentation, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/contributing.adoc

Additional resources

The OpenShift docs repo adheres to the following style guides:

- OpenShift documentation guidelines (OSDOCS)
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc
- Red Hat Supplementary Style Guide (SSG)
  https://redhat-documentation.github.io/supplementary-style-guide/
- Modular Documentation Reference Guide (Mod Docs)
  https://redhat-documentation.github.io/modular-docs/
- IBM Style Guide (ISG)
  https://www.ibm.com/docs/en/ibm-style

You can log in to the ISG by using your @redhat.com id and single sign-on (SSO) credentials. --->

lahinson commented:
Hi @sheriff-rh! As we discussed on Slack a few weeks ago, I've been working on the docs about configuring NTO in hosted control planes (aka HyperShift). When you can, would you mind taking a look at my draft? In particular, let me know if I got any of the NTO-related terminology wrong. Thank you! (Preview link is in the first comment 👆 )

lahinson commented:
@dagrayvid PTAL at the draft of the docs about configuring NTO in hosted control planes. The preview link is in the first comment 👆 

I also asked Andrew Taylor to take a look, since he usually covers the NTO-related docs. 

One question: Is Liquan Cui the right person from QE to review this? I'm not sure if I should ask the QE rep for PSAP or for HyperShift.

Thanks!

dagrayvid commented:
Thanks @lahinson for the draft. We are expanding on the NTO functionality in HyperShift a bit in [this PR](https://github.com/openshift/hypershift/pull/1729) which should be landing in the next hours / days (it has been fully approved). That PR includes new docs for the new features, so we should probably include/adapt those project docs into this PR as well.

lahinson commented:
Thanks for letting me know, @dagrayvid. I'll keep an eye on that PR, and after it's merged, I'll add the new info to this PR.

lahinson commented:
/assign lahinson

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+// * scalability_and_performance/using-node-tuning-operator.adoc
+
+[id="node-tuning-hosted-cluster_{context}"]
+= Configuring node tuning in a hosted cluster (Technology Preview)
```
No need for (Technology Preview) in the title, as that should be defined by the snippet below.

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+
+To set node-level tuning on the nodes in your hosted cluster, you can use the Node Tuning Operator. In hosted control planes, you can configure node tuning by creating config maps that contain tuned objects and referencing those config maps in your node pools. Currently, node tuning is limited to actions that the TuneD daemon can apply directly, such as setting sysctl values. Tuning that requires ...
+
+:FeatureName: Hosted control planes
```
The FeatureName and includes statement should be at the top.

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+.Procedure
+
+. Create a config map that contains a valid tuned manifest, and reference the manifest in a node pool. In the following example, a tuned manifest defines a profile that sets `vm.dirty_ratio` to 55 on nodes that contain the `tuned-1-node-label` node label with any value. Save the config map manifest in a file named `tuned-1.yaml`:
+
```
A couple things here. 
* Use a `+` sign to nest the code block beneath the step, as you have with the `[NOTE]` block. 
* For coloration on docs.openshift, include a [source,yaml] block. 
* You also need four dashes around the code block, before and after. 

Example: 
```
Save the config map manifest in a fila named `tuned-1.yaml`:
+
[source,yaml]
----
apiVersion:
etc...
----
+
[NOTE]
```

sheriff-rh replied:
So this is MUCH better, the build preview looks great! I noticed you still have empty lines between the step and the output on lines 18, 47, 62, 96, 113, and 122. While this still renders correctly on the docs.openshift build preview, there is no guarantee that it will render accurately on the customer portal or other places we publish.  I recommend removing empty lines where possible.

The second header and procedure needs to be in it's own module. Please trim the empty lines that are currently on lines  179, 194, 211, 229, 253, 270, and 290.

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+====
+
+. Create the config map in the management cluster:
+   
```
So this is a terminal command rather than yaml, which will look like this (using some of the same feedback from above). In addition, for terminal input we use a dollar sign to indicate user privileges. Sometimes we use a pound sign if different privileges are required (root users) but that's very infrequent in OpenShift:
```
. Create the config map in the management cluster: 
+
[source,terminal]
----
$ oc --kubeconfig .... .yaml
----
```

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+   oc --kubeconfig="$MGMT_KUBECONFIG" create -f tuned-1.yaml
+
+. Reference the config map in the node pool's `spec.tunedConfig` field, either by editing a node pool or creating one. In this example, assume that you have only one node pool, `nodepool-1`, which contains 2 nodes.
+
```
Same here, back to yaml block:
```
... 2 nodes:
+
[source,yaml]
----
apiVersion:
etc...
----
+
[NOTE]
```

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+Now that you have created the config map that contains a tuned manifest and referenced it in a node pool, the Node Tuning Operator syncs the tuned objects into the hosted cluster. You can verify which tuneds are defined and which profiles are set for each node.
+
+. List the tuned objects in the hosted cluster:
+
```
````
+
[source,terminal]
----
$ oc --kube...
----
````

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+
++
+.Example output
+[source,terminal]
```
Excellent! More of this :+1:

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+----
+   
+. List the profiles in the hosted cluster:
+
```
Here as well, terminal command

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+----
+
++
+As you can see, both worker nodes in the node pool have the `tuned-1-profile` applied. If no custom profiles are created, the `openshift-node` profile is applied by default.
```
The first sentence is unnecessary because the output shows the profile is applied. The second sentence should probably go in a [NOTE] block for emphasis. 

In addition, the `+` on the above line (107) needs to go directly under the output for it to nest correctly under the step. 
Example:
```
.Example output
[source,terminal]
----
    NAME                           TUNED            APPLIED   DEGRADED   AGE
    nodepool-1-worker-1            tuned-1-profile  True      False      7m43s
    nodepool-1-worker-2            tuned-1-profile  True      False      7m14s
----
+
[NOTE]
====
If no custom profiles are created, the `openshift-node` profile is applied by default.
===
```

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+As you can see, both worker nodes in the node pool have the `tuned-1-profile` applied. If no custom profiles are created, the `openshift-node` profile is applied by default.
+
+
+.Next steps
```
This should be `.Verification` because we are confirming the tuning was applied correctly.

sheriff-rh commented on scalability_and_performance/using-node-tuning-operator.adoc:
```diff
@@ -22,3 +22,10 @@ include::modules/custom-tuning-specification.adoc[leveloffset=+1]
+[role="_additional-resources"]
+.Additional resources
+
+For more information about hosted control planes, see link:https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.6/html/multicluster_engine/multicluster_engine_overview#hosted-control-planes-intro[Using hosted control plane clusters (Technology Preview)].
```
Is this documentation on docs.openshift? If so, we can do an `xref::` to the section.

lahinson replied:
Unfortunately, those docs aren't on docs.openshift. They're the docs for Red Hat Advanced Cluster Management.

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,121 @@
+.Next steps
+
+To confirm that the tuning was applied correctly, start a debug shell on a node and check the sysctl values:
+
```
```
+
[source,terminal]
----
$ oc kube...
----
```

sheriff-rh reviewed (changes requested):
This is a great start! I've made a few comments that I hope make it easier to understand how we indicate code blocks and nest them under steps correctly. Please reach out if you have any questions, or we can have a meeting and go through it together, too!

dagrayvid commented:
> Thanks for letting me know, @dagrayvid. I'll keep an eye on that PR, and after it's merged, I'll add the new info to this PR.

Thanks @lahinson, the PR has now merged!

lahinson commented:
@dagrayvid I updated this PR with the additional content. PTAL when you can.

@sheriff-rh I incorporated your feedback, and used what I learned to (hopefully) format the new content from David correctly. Let me know what you think.

@liqcui When you can, please take a look at this new content about node tuning for hosted control planes. The preview link is in the first comment in this PR. 

Thank you all!

liqcui commented:
/lgtm

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,296 @@
+====
+If no custom profiles are created, the `openshift-node` profile is applied by default.
+====
+
```
Remove one of these empty lines between the note and the next step.

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,296 @@
+    vm.dirty_ratio = 55
+----
+
+[id="advanced-node-tuning-hosted-cluster_{context}"]
```
One procedure per module where possible, according to the [guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#writing-procedures). I recommend breaking this out into a second module.

sheriff-rh reviewed (changes requested):
This looks much better! The build preview is improved, too. I left a couple comments with more detail, but there's three things to address: 
1. Remove empty lines above `+` to group them together. This occurs in several places. 
2. In steps and example outputs, remove the indent and add a `$` in front of `oc` steps. No dollar sign is needed for output steps. This occurs in most steps and outputs.
3. One procedure per module - Move the second procedure to it's own, new module and add that module `include::` statement to the assembly. 

As always - ping me or schedule a meeting if you want to go over anything, I'm happy to help!

lahinson commented:
Thanks a bunch, @sheriff-rh! I learned a lot from your feedback. I made the changes that you mentioned and squashed my commits. I also updated the preview link in the first comment so that you can jump directly to either of the two modules.

sheriff-rh commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+.Example output
+[source,terminal]
+----
+    NAME       AGE
```
One FINAL comment  - lines 97-100 are still indented. Everything else looks great!!

lahinson replied:
Good catch! I'll fix that now.

sheriff-rh reviewed (approved):
Excellent work! I'm on peer review this week, so I will go ahead and set the peer-review-done label on this PR. 

One example output needs the indent fixed, but otherwise it's ready to go! Great work!

sheriff-rh commented:
/label peer-review-done

lahinson commented:
Hi @dagrayvid- Any thoughts on the latest draft of the docs? https://51302--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/using-node-tuning-operator.html#node-tuning-hosted-cluster_node-tuning-operator

dagrayvid commented:
Hi @lahinson, I will try to do another review today or tomorrow... We do have one [pending change](https://github.com/openshift/hypershift/pull/1802#issue-1408554852) to rename a part of the NodePool spec which will impact the documentation, so I would suggest we hold this PR until that has merged.

lahinson commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+:FeatureName: Hosted control planes
+include::snippets/technology-preview.adoc[]
+
+To set node-level tuning on the nodes in your hosted cluster, you can use the Node Tuning Operator. In hosted control planes, you can configure node tuning by creating config maps that contain tuned objects and referencing those config maps in your node pools. Currently, node tuning is limited to actions that the TuneD daemon can apply directly, such as setting sysctl values. Tuning that requires ...
```
@dagrayvid Thanks for your review so far! I especially appreciate your thoughts on ConfigMap, NodePool, and Tuned. Would you say that all the instances in this content are referring to the actual objects? If so, I can update the terminology. 

The OpenShift glossary recommends "config map" and "node pool" unless the actual objects are being referred to, in which case, `ConfigMap` and `NodePool` are correct. I can't find guidance about Tuned, so I'll trust your advice that it should be capitalized.

dagrayvid replied:
I see what you mean. After taking a look through doc [guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting), it seems like some of the uses in these docs should be "config maps" and some should be `ConfigMap`... Specifically in this line, we are speaking in general so it is probably okay as it is.

dagrayvid commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+
+.Procedure
+
+. Create a config map that contains a valid tuned manifest, and reference the manifest in a node pool. In the following example, a tuned manifest defines a profile that sets `vm.dirty_ratio` to 55 on nodes that contain the `tuned-1-node-label` node label with any value. Save the config map manifest in a file named `tuned-1.yaml`:
```
From the guidelines, I think Tuned should be formatted here as `Tuned`, as we are talking about the specific example object.

dagrayvid replied:
```suggestion
. Create a config map that contains a valid tuned manifest, and reference the manifest in a node pool. In the following example, a `Tuned` manifest defines a profile that sets `vm.dirty_ratio` to 55 on nodes that contain the `tuned-1-node-label` node label with any value. Save the following `ConfigMap` manifest in a file named `tuned-1.yaml`:
```

dagrayvid commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+$ oc --kubeconfig="$MGMT_KUBECONFIG" create -f tuned-1.yaml
+----
+
+. Reference the config map in the node pool's `spec.tunedConfig` field, either by editing a node pool or creating one. In this example, assume that you have only one node pool, `nodepool-1`, which contains 2 nodes.
```
```suggestion
. Reference the `ConfigMap` in the node pool's `spec.tunedConfig` field, either by editing a node pool or creating one. In this example, assume that you have only one `NodePool`, `nodepool-1`, which contains 2 nodes.
```

dagrayvid commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+
+.Verification
+
+Now that you have created the config map that contains a tuned manifest and referenced it in a node pool, the Node Tuning Operator syncs the tuned objects into the hosted cluster. You can verify which tuneds are defined and which profiles are set for each node.
```
```suggestion
Now that you have created the `ConfigMap` that contains a `Tuned` manifest and referenced it in a `NodePool`, the Node Tuning Operator syncs the `Tuned` objects into the hosted cluster. You can verify which `Tuned` objects are defined and which TuneD profiles are applied to each node.
```

dagrayvid commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+
+Now that you have created the config map that contains a tuned manifest and referenced it in a node pool, the Node Tuning Operator syncs the tuned objects into the hosted cluster. You can verify which tuneds are defined and which profiles are set for each node.
+
+. List the tuned objects in the hosted cluster:
```
```suggestion
. List the `Tuned` objects in the hosted cluster:
```

dagrayvid commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+tuned-1    65s
+----
+   
+. List the profiles in the hosted cluster:
```
```suggestion
. List the `Profile` objects in the hosted cluster:
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+
+.Procedure
+
+. Create a config map that contains a tuned object manifest for creating 10 huge pages that are 2 MB in size. Save this config map manifest in a file named `tuned-hugepages.yaml`:
```
```suggestion
. Create a `ConfigMap` that contains a `Tuned` object manifest for creating 10 huge pages that are 2 MB in size. Save this `ConfigMap` manifest in a file named `tuned-hugepages.yaml`:
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
++
+[NOTE]
+====
+The `.spec.recommend.match` field is intentionally left blank. In this case, this tuned is applied to all nodes in the node pool where this config map is referenced. Group nodes with the same hardware configuration into the same node pool. Otherwise, TuneD operands can calculate conflicting kernel parameters for two or more nodes that share the same node pool.
```
```suggestion
The `.spec.recommend.match` field is intentionally left blank. In this case, this `Tuned` is applied to all nodes in the node pool where this `ConfigMap` is referenced. Group nodes with the same hardware configuration into the same node pool. Otherwise, TuneD operands can calculate conflicting kernel parameters for two or more nodes that share the same node pool.
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+The `.spec.recommend.match` field is intentionally left blank. In this case, this tuned is applied to all nodes in the node pool where this config map is referenced. Group nodes with the same hardware configuration into the same node pool. Otherwise, TuneD operands can calculate conflicting kernel parameters for two or more nodes that share the same node pool.
+====
+
+. Create the config map in the management cluster:
```
```suggestion
. Create the `ConfigMap` in the management cluster:
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+$ oc --kubeconfig="$MGMT_KUBECONFIG" create -f tuned-hugepages.yaml
+----
+
+. Create a node pool manifest YAML file, customize the upgrade type of the node pool, and reference the config map that you created in the `spec.tunedConfig` section. Create the node pool manifest and save it in a file named `hugepages-nodepool.yaml`:
```
```suggestion
. Create a `NodePool` manifest YAML file, customize the upgrade type of the `NodePool` and reference the `Config Map` that you created in the `spec.tunedConfig` section. Create the `NodePool` manifest and save it in a file named `hugepages-nodepool.yaml` using the `hypershift` CLI:
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+      --render > hugepages-nodepool.yaml
+----
+
+. In the `hugepages-nodepool.yaml` file, set `.spec.management.upgradeType` to `InPlace`, and set `.spec.tunedConfig` to reference the `tuned-hugepages` config map that you created.
```
```suggestion
. In the `hugepages-nodepool.yaml` file, set `.spec.management.upgradeType` to `InPlace`, and set `.spec.tunedConfig` to reference the `tuned-hugepages` ConfigMap that you created.
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
++
+[NOTE]
+====
+To avoid the unnecessary re-creation of nodes when you apply the new 'MachineConfig' objects, set `.spec.management.upgradeType` to `InPlace`. If you use the `Replace` upgrade type, nodes are fully deleted and new nodes can replace them when you apply the new kernel boot parameters that the TunedD operand calculated.
```
```suggestion
To avoid the unnecessary re-creation of nodes when you apply the new 'MachineConfig' objects, set `.spec.management.upgradeType` to `InPlace`. If you use the `Replace` upgrade type, nodes are fully deleted and new nodes can replace them when you apply the new kernel boot parameters that the TuneD operand calculated.
```

dagrayvid replied:
@lahinson I think this typo is still in the docs.

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+To avoid the unnecessary re-creation of nodes when you apply the new 'MachineConfig' objects, set `.spec.management.upgradeType` to `InPlace`. If you use the `Replace` upgrade type, nodes are fully deleted and new nodes can replace them when you apply the new kernel boot parameters that the TunedD operand calculated.
+====
+
+. Create the node pool in the management cluster:
```
```suggestion
. Create the `NodePool` in the management cluster:
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+rendered             123m
+----
+   
+. List the profiles in the hosted cluster:
```
```suggestion
. List the `Profile` objects in the hosted cluster:
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+
+After the nodes are available, the containerized TuneD daemon calculates the required kernel boot parameters based on the applied tuned profile. After the nodes are ready and reboot once to apply the generated 'MachineConfig' object, you can verify that the tuned profile is applied and that the kernel boot parameters are set.
+
+. List the tuned objects in the hosted cluster:
```
```suggestion
. List the `Tuned` objects in the hosted cluster:
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+hugepages-nodepool-worker-2    openshift-node-hugepages   True      False      3m57s
+----
++
+Both of the worker nodes in the new node pool have the `openshift-node-hugepages` profile applied.
```
```suggestion
Both of the worker nodes in the new `NodePool` have the `openshift-node-hugepages` profile applied.
```

dagrayvid commented:
Thanks @lahinson, based on the guidelines I went through the changes and made my best attempt at deciding the correct formatting for the different object names (Tuned, NodePool, ConfigMap, etc...), though I am open to a second opinion on these!

I also made some other minor suggestions.

lahinson commented:
Thanks so much, @dagrayvid! I'll go through your revisions and will let you know if I have any questions.

lahinson commented:
@dagrayvid I incorporated your feedback, and you can see the latest preview here: https://51302--docspreview.netlify.app/openshift-enterprise/latest/scalability_and_performance/using-node-tuning-operator.html#node-tuning-hosted-cluster_node-tuning-operator

PTAL when you can. Thanks!

dagrayvid commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+$ oc --kubeconfig="$MGMT_KUBECONFIG" create -f tuned-1.yaml
+----
+
+. Reference the `ConfigMap` object in the `spec.tunedConfig` field of the node pool, either by editing a node pool or creating one. In this example, assume that you have only one 'NodePool', named `nodepool-1`, which contains 2 nodes.
```
```suggestion
. Reference the `ConfigMap` object in the `spec.tunedConfig` field of the node pool, either by editing a node pool or creating one. In this example, assume that you have only one `NodePool`, named `nodepool-1`, which contains 2 nodes.
```

dagrayvid replied:
```suggestion
. Reference the `ConfigMap` object in the `spec.tuningConfig` field of the node pool, either by editing a node pool or creating one. In this example, assume that you have only one 'NodePool', named `nodepool-1`, which contains 2 nodes.
```

dagrayvid reviewed (commented):
Thanks @lahinson, the changes look good to me. I just found one instance where ' was used instead of `.

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+$ oc --kubeconfig="$MGMT_KUBECONFIG" create -f tuned-hugepages.yaml
+----
+
+. Create a `NodePool` manifest YAML file, customize the upgrade type of the `NodePool`, and reference the `ConfigMap` object that you created in the `spec.tunedConfig` section. Create the `NodePool` manifest and save it in a file named `hugepages-nodepool.yaml` by using the `hypershift` CLI:
```
```suggestion
. Create a `NodePool` manifest YAML file, customize the upgrade type of the `NodePool`, and reference the `ConfigMap` object that you created in the `spec.tuningConfig` section. Create the `NodePool` manifest and save it in a file named `hugepages-nodepool.yaml` by using the `hypershift` CLI:
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+      name: tuned-hugepages
+      namespace: clusters
+    data:
+      tuned: |
```
```suggestion
      tuning: |
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+      --render > hugepages-nodepool.yaml
+----
+
+. In the `hugepages-nodepool.yaml` file, set `.spec.management.upgradeType` to `InPlace`, and set `.spec.tunedConfig` to reference the `tuned-hugepages` `ConfigMap` object that you created.
```
```suggestion
. In the `hugepages-nodepool.yaml` file, set `.spec.management.upgradeType` to `InPlace`, and set `.spec.tuningConfig` to reference the `tuned-hugepages` `ConfigMap` object that you created.
```

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+        ...
+        upgradeType: InPlace
+      ...
+      tunedConfig:
```
```suggestion
      tuningConfig:
```

dagrayvid commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+      name: tuned-1
+      namespace: clusters
+    data:
+      tuned: |
```
```suggestion
      tuning: |
```

dagrayvid commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+    ...
+    spec:
+      ...
+      tunedConfig:
```
```suggestion
      tuningConfig:
```

dagrayvid reviewed (commented):
I am adding an additional review to summarize the changes that will be needed **after** https://github.com/openshift/hypershift/pull/1802 merges. We are renaming a part of the API. These changes should be similar to the docs changes in the PR for the upstream HyperShift project docs.

lahinson commented:
@dagrayvid Thanks for your additional review! I made the changes that you suggested. Do you think we're at a good place for me to move this forward to be merged?

dagrayvid commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
+
+.Verification
+
+After the nodes are available, the containerized TuneD daemon calculates the required kernel boot parameters based on the applied TuneD profile. After the nodes are ready and reboot once to apply the generated 'MachineConfig' object, you can verify that the TuneD profile is applied and that the kernel boot parameters are set.
```
```suggestion
After the nodes are available, the containerized TuneD daemon calculates the required kernel boot parameters based on the applied TuneD profile. After the nodes are ready and reboot once to apply the generated `MachineConfig` object, you can verify that the TuneD profile is applied and that the kernel boot parameters are set.
```

dagrayvid commented on modules/node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,134 @@
+$ oc --kubeconfig="$MGMT_KUBECONFIG" create -f tuned-1.yaml
+----
+
+. Reference the `ConfigMap` object in the `spec.tuningConfig` field of the node pool, either by editing a node pool or creating one. In this example, assume that you have only one 'NodePool', named `nodepool-1`, which contains 2 nodes.
```
```suggestion
. Reference the `ConfigMap` object in the `spec.tuningConfig` field of the node pool, either by editing a node pool or creating one. In this example, assume that you have only one `NodePool`, named `nodepool-1`, which contains 2 nodes.
```

lahinson commented on modules/advanced-node-tuning-hosted-cluster.adoc:
```diff
@@ -0,0 +1,154 @@
++
+[NOTE]
+====
+To avoid the unnecessary re-creation of nodes when you apply the new 'MachineConfig' objects, set `.spec.management.upgradeType` to `InPlace`. If you use the `Replace` upgrade type, nodes are fully deleted and new nodes can replace them when you apply the new kernel boot parameters that the TuneD operand calculated.
```
Good catch, @dagrayvid! I'll be sure to fix those.

lahinson commented:
/merge-review-needed

jeana-redhat reviewed (commented):
The modules should have content type metadata, once that's fixed this can be merged.

/remove-label merge-review-in-progress
/remove-label merge-review-needed

lahinson commented:
Good catch, @jeana-redhat. I added the metadata. Thanks!

/label merge-review-needed

jeana-redhat commented:
Great! will merge once Travis is done. Thanks :)

/remove-label merge-review-in-progress
/remove-label merge-review-needed

jeana-redhat commented:
/cherrypick enterprise-4.12

jeana-redhat commented:
hm :thinking: 
/cherrypick enterprise-4.12
